debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/StreamingWal2JsonMessageDecoder.java

```diff
@@ -0,0 +1,170 @@
+ *          "columnvalues": [1, 1.10, 10, 10.1111, 3.30, 22.2200, 22, 22.2222, 1.00, 10, 10, 22.0000, 22, 22, null, null, null, null, null, null]
+ *      }
+ * </pre>
+ *
+ * <b>Further changes</b>
+ * <pre>
+ *      ,{
+ *          "kind": "insert",
+ *          "schema": "public",
+ *          "table": "numeric_decimal_table",
+ *          "columnnames": ["pk", "d", "dzs", "dvs", "d_nn", "n", "nzs", "nvs", "d_int", "dzs_int", "dvs_int", "n_int", "nzs_int", "nvs_int", "d_nan", "dzs_nan", "dvs_nan", "n_nan", "nzs_nan", "nvs_nan"],
+ *          "columntypes": ["integer", "numeric(3,2)", "numeric(4,0)", "numeric", "numeric(3,2)", "numeric(6,4)", "numeric(4,0)", "numeric", "numeric(3,2)", "numeric(4,0)", "numeric", "numeric(6,4)", "numeric(4,0)", "numeric", "numeric(3,2)", "numeric(4,0)", "numeric", "numeric(6,4)", "numeric(4,0)", "numeric"],
+ *          "columnoptionals": [false, true, true, true, false, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true],
+ *          "columnvalues": [1, 1.10, 10, 10.1111, 3.30, 22.2200, 22, 22.2222, 1.00, 10, 10, 22.0000, 22, 22, null, null, null, null, null, null]
+ *      }
+ * </pre>
+ *
+ * <b>End of message</b>
+ * <pre>
+ *      ]
+ * }
+ * </pre>
+ * </p>
+ *
+ * <p>
+ * For parsing purposes it is necessary to add or remove a fragment of JSON to make a well-formatted JSON out of it.
+ * The last message is just dropped.
+ * </p>
+ * @author Jiri Pechanec
+ *
+ */
+public class StreamingWal2JsonMessageDecoder implements MessageDecoder {
+
+    private static final  Logger LOGGER = LoggerFactory.getLogger(StreamingWal2JsonMessageDecoder.class);
+
+    private final DateTimeFormat dateTime = DateTimeFormat.get();
+    private boolean containsMetadata = false;
+    private boolean messageInProgress = false;
+    private String bufferedContent;
+
+    private int txId;
+
+    private String timestamp;
+
+    private long commitTime;
+
+    @Override
+    public void processMessage(ByteBuffer buffer, ReplicationMessageProcessor processor, TypeRegistry typeRegistry) throws SQLException, InterruptedException {
+        try {
+            if (!buffer.hasArray()) {
+                throw new IllegalStateException("Invalid buffer received from PG server during streaming replication");
+            }
+            final byte[] source = buffer.array();
+            String content = new String(Arrays.copyOfRange(source, buffer.arrayOffset(), source.length)).trim();
+            LOGGER.debug("Chunk arrived from database {}", content);
+
+            if (!messageInProgress) {
+                // We received the beginning of a transaction
+                if (!content.endsWith("}")) {
+                    // Chunks are enabled and we have an unfinished message, it is necessary to add a sequence of closing chars
+                    content += "]}";
+                }
+                final Document message = DocumentReader.defaultReader().read(content);
```

Comment:
`DocumentReader` has a `read(byte[])` method, too. It seems we could avoid the conversion into a String completely? Checking for specific begin/end characters would be a tad less elegant as we'd have to check the byte at the given position, but it seems worth it to avoid the overhead.

Reply:
@gunnarmorling And tend to disagrre with this - can we guarantee there are not UTF-8 chars in the message? If not then we should recode byte[] to char[] and in that case the `String` encapsulation just adds only a small overhead. We might optimize a bit by turning `content` into a `StringBuilder` and use `charAt` for testing

Reply:
Good point. We should clarify what's the encoding of incoming messages. That being said, the existing implementation (now in `NonStreamingWal2JsonMessageDecoder`) uses the `read(byte[])` method. So it seems things at least wouldn't get worse when doing the same here (arguably might be better than relying on the default encoding of the platform).

Reply:
@gunnarmorling I've dropped all string operations and use `byte[]` only